### ISSUE TITLE
Mismatched types error E0308 when there is no error

### ISSUE THREAD

**Issue by supercurio** [intellij-rust/intellij-rust#1753]
With the new plugin 52, Version: 0.1.0.2069 and using the [jni-rs crate](https://crates.io/crates/jni), this error is reported at many places in a Rust / JNI codebase.

```rust
JObject::from(env.new_string("hello")?);
```

> mismatched types E0308
> expected JThrowable, found JString

```rust
JString::from(elem_obj)
```

> mismatched types E0308
> expected *mut _jobject, found JObject

JObject and JString can be converted to each other using ::from.
source: https://github.com/prevoty/jni-rs/blob/master/src/wrapper/objects/jstring.rs#L25

In the meantime, would there be an way to disable this check? It persists after disabling all inspections in IDEA Settings.
Thanks!

**Comment by Zteve** [comment]
I'm seeing this, too, in my code and in the std libraries when browsing them in IntelliJ-Rust. Latest version of everything (Plugin: 0.1.0.2069; IntelliJ: 2017.2.4; Rustc: 1.20.0).

Simple example:
```
pub fn validate<S>(input: S) -> Result<LodaModel, GeneralError>
    where S: Into<String>
{
    let open_file = fs::File::open(PathBuf::from(input.into()))?;

    LodaModel::from_reader(open_file).map_err(|generr| From::from(generr))
}
```
Complains about this line:
```
    let open_file = fs::File::open(PathBuf::from(input.into()))?;
                                                 ~~~~~~~~~~~~
```
with message: _mismatched types E0308 expected Box\<Path\>, found String_.

Compiles (from shell command `cargo build`) with no warnings or errors.

**Comment by andy-hanson** [comment]
I too have an error related to a conversion not happening:
```rust
use std::ops::Deref;

struct Wrapper<T>(T);
impl<T> Deref for Wrapper<T> {
	type Target = T;
	fn deref(&self) -> &T {
		&self.0
	}
}

struct RefWrapper<'a, T : 'a>(&'a T);

struct WrapperWrapper<'a>(RefWrapper<'a, u32>);

fn f(w: &Wrapper<u32>) -> WrapperWrapper {
	WrapperWrapper(RefWrapper(w))
}
```

In the command-line compiler there is no error. In IntelliJ, I get an error at `RefWrapper(w)`:
```
mismatched types E0308
expected RefWrapper<u32>, found RefWrapper<Wrapper<u32>>
```

(Aside: I ran into https://youtrack.jetbrains.com/issue/IDEA-65636 and had to copy the error message from memory.)

**Comment by Zteve** [comment]
@andy-hanson Re your aside: this "bug"/"omission" in IntelliJ IDEA has a long history (including tooltip copying) and seems to have stabilised on a single (non-closed) issue [IDEA-168283](https://youtrack.jetbrains.com/issue/IDEA-168283) which talks about error messages.  There seems to be reluctance to offer copy functionality (easily) to these dynamic message boxes—though I cannot see why they simply don't allow us to 'pin' them.

My solution is to (carefully) select the text, and, hit Cmd+C (Ctrl+C on non macOS systems) while holding the mouse button down.  It is fiddly, since the box disappears if you're not careful, but I can often copy (most of) the text this way.

**Comment by andy-hanson** [comment]
Found another one:
```rust
struct Res(Option<bool>);
fn sum_optional_point(o: Option<u32>) -> Res {
	match o {
		Some(_) => Res(None),
		None => Res(None),
	}
}
```

At the second `Res(None)`, it gives me:
```
mismatched types E0308
expected Option<bool>, found Option<u32>
```

@Zteve thanks for the help.

**Comment by Techcable** [comment]
There seems to be some strange combination with the issue causing the inference algorithm to misunderstand of return types. With the method signature `pub fn alloc<T: HirContextAllocated>(&self, value: T) -> &T::Allocated`. It infers a ridiculous return type incorrectly after I invoke `HirInstructionVarargs::from`. 
![](https://i.gyazo.com/383371db287d8d999e0fdf12e7ab7f18.png)
The `fn(SmallVec<[HirVariable; 8]>) -> HirInstructionVarargs` part needs to be replaced with just `HirInstructionVarargs`, as it seems the type inference will work after that.
Excellent work on getting type inference this far, @vlad20012 is a wizard!

**Comment by vlad20012** [comment]
@matklad, take a look to @andy-hanson's case. There is a name resolution failure: here `None => Res(None),` this `None` resolves locally to the `None` in pattern.

**Comment by vlad20012** [comment]
@Techcable the real wizard is @matklad, who patiently created the infrastructure for 2 years, using which I did it in 1-2 lines of code =)
You can become a "wizard" like me, just start contributing here!

**Comment by aoprisan** [comment]
I guess I am experiencing the same issue  with ``` to_string```  being called on ```&str```, like in the following example:
```
ReportService::temp_report_location(filename, "html".to_string());
```
The error is: ```mismatched types E0380: expecting String, found String```
![screenshot](https://user-images.githubusercontent.com/99908/31088732-6591d8c2-a7aa-11e7-95ae-b33bd530b3f0.png)

Rust plugin version: Version: 0.1.0.2070
IDEA Community Edition EAP 2017.3 : IC-173.2969.26

**Comment by cmaher** [comment]
I'm seeing this with 

```rust
let guess: u32 = match guess.trim().parse() {
            Ok(num) => num,
            Err(_) => {
                println!("Not a number!");
                continue;
            }
};
```

As a workaround for now, I've set highlighting level to "None" and invalidated caches + restarted.

**Comment by vlad20012** [comment]
We've disabled typechecking by default in the latest release #1836 due to these issues :(

**Comment by supercurio** [comment]
Thanks, the false positive were making this feature counter-productive. 
In the meantime, the cargo check option does the job.

Looking forward to the typechecking capability return once it's reliable enough!

**Comment by vlad20012** [comment]
The issue is still actual for disabled-by-default inspection =)
